h5py/_hl/dataset.py

```diff
@@ -213,10 +214,16 @@ def __len__(self):
         """
         return len(self._dset)
 
-    def __array__(self, dtype=None):
+    def __array__(self, dtype=None, copy=True):
+        if copy not in (True, _NUMPY_COPY_IF_NEEDED):
+            raise ValueError(
+                f"AstypeWrapper.__array__ received {copy=} "
+                f"but memory allocation cannot be avoided on read"
+            )
+
         data = self[:]
         if dtype is not None:
-            data = data.astype(dtype)
+            return data.astype(dtype, copy=_NUMPY_COPY_IF_NEEDED)
```

Comment:
```suggestion
            return data.astype(dtype, copy=False)
```

I'm pretty sure this will work on both NumPy versions. `data` at this point is a newly allocated array from reading the data.

[array.astype() docs for Numpy 2.x](https://numpy.org/devdocs/reference/generated/numpy.ndarray.astype.html#numpy.ndarray.astype) seem to confirm this should still work.